### PR TITLE
[BC Break] Add ReferenceOneField

### PR DESCRIPTION
**Pull request by fzaninotto**
## Problem

I started documenting the react-admin ReferenceXXXfFields, and found ReferenceOne to be a missing spot in our strategy.

## Solution

- [x] Remove props injection in ReferenceField
- [x] Add ReferenceOneField (and the `useReferenceOneFieldController`), reusing the view from `<ReferenceField>`
- [x] Add documentation for relationship types and the react-admin solutions to handle them

### CODE REVIEW

**Review by djhi — APPROVED** [review]

**Comment by WiXSL** [comment]
Reviewing...

File: docs/FieldsForRelationships.md [hunk 1]
```diff
@@ -0,0 +1,414 @@
...
+
+To fetch the books of an author with react-admin, you can use:
+
+- [`<ReferenceManyField>`](#referencemanyfield) when the API uses a foreign key (e.g. each book has an `author_id` field)
+- [`<ReferenceArrayField>`](#referencearrayfield) when the API uses an array of foreign keys (e.g. each author has a `book_ids` field)
+- [`<ArrayField>`](#arrayfield) when the API embeds an array of records (e.g. each author has an `books` field)
+
+## Many-To-One
+
+On the other hand, **many-to-one relationships** are the opposite of one-to-many relationships (e.g. each book has one author). To fetch the the author of a book, you can use:
```
Comment (WiXSL) [inline docs/FieldsForRelationships.md:31]:
```suggestion
On the other hand, **many-to-one relationships** are the opposite of one-to-many relationships (e.g. each book has one author). To fetch the author of a book, you can use:
```

File: docs/FieldsForRelationships.md [hunk 2]
```diff
@@ -0,0 +1,414 @@
...
+        </SimpleShowLayout>
+    </Show>
+);
+```
+
+`<ReferenceField>` uses the current `record` (a book in this example) to read the id of the reference using the foreign key (`author_id`). Then, it uses `dataProvider.getOne('authors', { id })` fetch the related author.
+
+`<ReferenceField>` creates a `RecordContext` with the reference record, so you can use any component relying on this context (`<TextField>`, `<SimpleShowLayout>`, etc.).
+
+**Tip**: You don't need to worry about the fact that this components calls `<ReferenceField>` twice on the same table. React-admin will only make one call to the API.
```
Comment (WiXSL) [inline docs/FieldsForRelationships.md:170]:
```suggestion
**Tip**: You don't need to worry about the fact that these components calls `<ReferenceField>` twice on the same table. React-admin will only make one call to the API.
```

**Review by WiXSL — CHANGES_REQUESTED** [review]

File: docs/FieldsForRelationships.md [hunk 3]
```diff
@@ -0,0 +1,414 @@
...
+
+## One-To-Many
+
+When one record has many related records, this is called a one-to-many relationship. For instance, if an author has written several books, `authors` has a one-to-many relationship with `books`. 
+
+To fetch the books of an author with react-admin, you can use:
+
+- [`<ReferenceManyField>`](#referencemanyfield) when the API uses a foreign key (e.g. each book has an `author_id` field)
+- [`<ReferenceArrayField>`](#referencearrayfield) when the API uses an array of foreign keys (e.g. each author has a `book_ids` field)
+- [`<ArrayField>`](#arrayfield) when the API embeds an array of records (e.g. each author has an `books` field)
```
Comment (WiXSL) [inline docs/FieldsForRelationships.md:27]:
```suggestion
- [`<ArrayField>`](#arrayfield) when the API embeds an array of records (e.g. each author has a `books` field)
```

**Review by WiXSL — CHANGES_REQUESTED** [review]

File: docs/FieldsForRelationships.md [hunk 4]
```diff
@@ -0,0 +1,414 @@
...
+On the other hand, **many-to-one relationships** are the opposite of one-to-many relationships (e.g. each book has one author). To fetch the the author of a book, you can use:
+
+- [`<ReferenceField>`](#referencefield) when the API uses a foreign key (e.g. each book has an `author_id` field)
+- [Deep Field Source](#deep-field-source), when the API embeds the related record (e.g. each book has an `author` field containing an object)
+
+Other kinds of relationships often reduce to one-to-many relationships. 
+
+## One-To-One
+
+For instance, **one-to-one relationships** (e.g. a book has one book_detail) are a special type of one-to-many relationship with a cardinality of 1. To fetch the details of a book, you can use:
```
Comment (WiXSL) [inline docs/FieldsForRelationships.md:40]:
```suggestion
For instance, **one-to-one relationships** (e.g. a book has one `book_detail`) are a special type of one-to-many relationship with a cardinality of 1. To fetch the details of a book, you can use:
```

**Review by WiXSL — CHANGES_REQUESTED** [review]

File: docs/FieldsForRelationships.md [hunk 5]
```diff
@@ -0,0 +1,414 @@
...
+        </SimpleShowLayout>
+    </Show>
+);
+```
+
+`<ReferenceManyField>` uses the current `record` (an author in this example) to build a filter for the list of books on the foreign key field (`author_id`). Then, it uses `dataProvider.getManyReference('books', { target: 'author_id', id: book.id })` fetch the related books.
+
+`<ReferenceManyField>` creates a `ListContext` with the related records, so you can use any component relying on this context (`<Datagrid>`, `<SimpleList>`, etc.).
+
+**Tip**: For many APIs, there is no difference between `dataProvider.getList()` and `dataProvider.getManyReference()`. The latter is a specialized version of the former, with a predefined `filter`. But some APIs expose related records as a subroute, and therefore need a special method to fetch them. For instance, the boks of an author can be exposed via the following endpoint:
```
Comment (WiXSL) [inline docs/FieldsForRelationships.md:234]:
```suggestion
**Tip**: For many APIs, there is no difference between `dataProvider.getList()` and `dataProvider.getManyReference()`. The latter is a specialized version of the former, with a predefined `filter`. But some APIs expose related records as a sub-route, and therefore need a special method to fetch them. For instance, the books of an author can be exposed via the following endpoint: 
```

**Review by WiXSL — CHANGES_REQUESTED** [review]

File: docs/FieldsForRelationships.md [hunk 6]
```diff
@@ -0,0 +1,414 @@
...
+            <ReferenceOneField label="ISBN" reference="book_details" target="book_id">
+                <TextField source="ISBN" />
+            </ReferenceManyToManyField>
+            <EditButton />
+        </SimpleShowLayout>
+    </Show>
+);
+```
+
+`<ReferenceOneField>` behaves like `<ReferenceManyField>`: it uses the current `record` (a book in this example) to build a filter for the book details with the same the foreign key (`book_id`). Then, it uses `dataProvider.getManyReference('book_details', { target: 'book_id', id: book.id })` fetch the related details, and takes the first one.
```
Comment (WiXSL) [inline docs/FieldsForRelationships.md:408]:
```suggestion
`<ReferenceOneField>` behaves like `<ReferenceManyField>`: it uses the current `record` (a book in this example) to build a filter for the book details with the foreign key (`book_id`). Then, it uses `dataProvider.getManyReference('book_details', { target: 'book_id', id: book.id })` to fetch the related details, and takes the first one.
```

File: docs/FieldsForRelationships.md [hunk 7]
```diff
@@ -0,0 +1,414 @@
...
+
+`<ReferenceOneField>` creates a `RecordContext` with the reference record, so you can use any component relying on this context (`<TextField>`, `<SimpleShowLayout>`, etc.).
+
+**Tip**: As with `<ReferenceField>`, you can call `<ReferenceOneField>` as many times as you need in the same component, react-admin will only make one call to `dataProvider.getManyReference()`.
```
Comment (WiXSL) [inline docs/FieldsForRelationships.md:412]:
```suggestion
**Tip**: As with `<ReferenceManyField>`, you can call `<ReferenceOneField>` as many times as you need in the same component, react-admin will only make one call to `dataProvider.getManyReference()`.
```

**Review by WiXSL — CHANGES_REQUESTED** [review]

**Review by WiXSL — CHANGES_REQUESTED** [review]
`ColoredNumberField` component in the demo is also using `record` as a prop instead of taking it from the context

File: packages/ra-ui-materialui/src/field/ReferenceOneField.tsx [hunk 8]
```diff
@@ -0,0 +1,108 @@
...
+    return !record || (!isLoading && referenceRecord == null) ? (
+        emptyText ? (
+            <Typography component="span" variant="body2">
+                {emptyText}
+            </Typography>
+        ) : null
+    ) : (
+        <ResourceContextProvider value={reference}>
+            <ReferenceFieldView
+                isLoading={isLoading}
```
Comment (WiXSL) [inline packages/ra-ui-materialui/src/field/ReferenceOneField.tsx:69]: I get a typescript error here: 
`Property 'isLoading' does not exist on type 'IntrinsicAttributes & ReferenceFieldViewProps & { children?: ReactNode; }'.`

File: packages/ra-ui-materialui/src/field/ReferenceField.tsx [hunk 9]
```diff
@@ -169,21 +167,12 @@ const stopPropagation = e => e.stopPropagation();
 
 export const ReferenceFieldView: FC<ReferenceFieldViewProps> = props => {
     const {
-        basePath,
         children,
         className,
         error,
```
Comment (WiXSL) [inline packages/ra-ui-materialui/src/field/ReferenceField.tsx:172]: I get Typescript errors:

`TS2339: Property 'error' does not exist on type 'ReferenceFieldViewProps & { children?: ReactNode; }'.`
`TS2339: Property 'isLoading' does not exist on type 'ReferenceFieldViewProps & { children?: ReactNode; }'.`
`TS2339: Property 'referenceRecord' does not exist on type 'ReferenceFieldViewProps & { children?: ReactNode; }'.`

**Review by WiXSL — CHANGES_REQUESTED** [review]

**Comment by WiXSL** [comment]
Finish reviewing.
The new documentation explaining in detail the relationships fields looks great!

File: packages/ra-ui-materialui/src/field/ReferenceField.tsx (hunk 9 above)
Comment (fzaninotto) [reply packages/ra-ui-materialui/src/field/ReferenceField.tsx:172]: Weird. I don't see it and neither does the TypeScript compilation step in the CI. Did you rebuild the ra-core types?

File: packages/ra-ui-materialui/src/field/ReferenceOneField.tsx (hunk 8 above)
Comment (fzaninotto) [reply packages/ra-ui-materialui/src/field/ReferenceOneField.tsx:69]: same

**Review by WiXSL — APPROVED** [review]

File: packages/ra-ui-materialui/src/field/ReferenceField.tsx (hunk 9 above)
Comment (WiXSL) [reply packages/ra-ui-materialui/src/field/ReferenceField.tsx:172]: You are right. I re-build everything and it works

**Pull request merged**